src/viam/sdk/spatialmath/orientation.cpp

```diff
@@ -14,158 +13,144 @@
-        }
-        case proto::Orientation::TypeCase::kQuaternion: {
-            cfg.type_ = Quaternion;
-            quaternion quat;
-            quat.w = proto.quaternion().w();
-            quat.x = proto.quaternion().x();
-            quat.y = proto.quaternion().y();
-            quat.z = proto.quaternion().z();
-            cfg.orientation_ = quat;
-            break;
-        }
-        case proto::Orientation::TypeCase::kVectorDegrees: {
-            cfg.type_ = OrientationVectorDegrees;
-            orientation_vector_degrees ovd;
-            ovd.x = proto.vector_degrees().x();
-            ovd.y = proto.vector_degrees().y();
-            ovd.z = proto.vector_degrees().z();
-            ovd.theta = proto.vector_degrees().theta();
-            cfg.orientation_ = ovd;
-            break;
-        }
-        case proto::Orientation::TypeCase::kVectorRadians: {
-            cfg.type_ = OrientationVector;
-            orientation_vector ov;
-            ov.x = proto.vector_radians().x();
-            ov.y = proto.vector_radians().y();
-            ov.z = proto.vector_radians().z();
-            ov.theta = proto.vector_radians().theta();
-            cfg.orientation_ = ov;
-            break;
-        }
-        case proto::Orientation::TypeCase::kNoOrientation: {
-            // if type is NoOrientation, we put a sentinel
-            // orientation that indicates no rotation
-            cfg.type_ = Quaternion;
-            quaternion quat;
-            quat.w = 1;
-            quat.x = 0;
-            quat.y = 0;
-            quat.z = 0;
-            cfg.orientation_ = quat;
-            break;
+        auto operator()(const orientation_vector&) const {
+            return OrientationType::OrientationVector;
         }
-        case proto::Orientation::TypeCase::TYPE_NOT_SET:
-        default: {
-            throw Exception(ErrorCondition::k_not_supported, "orientation type not known");
+
+        auto operator()(const orientation_vector_degrees&) const {
+            return OrientationType::OrientationVectorDegrees;
         }
-    }
-    return cfg;
+
+        auto operator()(const euler_angles&) const {
+            return OrientationType::EulerAngles;
+        }
+    };
+
+    return boost::apply_visitor(Visitor{}, o);
 }
 
-proto::Orientation OrientationConfig::to_proto() const {
-    proto::Orientation orientation;
-    switch (type_) {
-        case AxisAngles: {
-            proto::Orientation_AxisAngles aa;
-            const axis_angles a = boost::get<axis_angles>(orientation_);
+namespace proto_convert_details {
+
+void to_proto<Orientation>::operator()(const Orientation& self, app::v1::Orientation* o) const {
+    struct Visitor {
+        void operator()(const axis_angles& a) {
+            app::v1::Orientation_AxisAngles aa;
             aa.set_x(a.x);
             aa.set_y(a.y);
             aa.set_z(a.z);
             aa.set_theta(a.theta);
             *orientation.mutable_axis_angles() = std::move(aa);
```

Comment:
(q) I think I must be misunderstanding something, how does this compile? Where is `orientation` defined? It looks to me like the proto `Orientation` value has been renamed `o`, from `orientation`. What am I missing?

edit: it looks like `orientation` is being defined after the `Visitor` struct, is that correct? I'm a little bit surprised that's allowed in general, but also I wonder if I'm interpreting that correctly because it seems to me we should be modifying the passed-in `o` value, not the `orientation` value created on line 90.

Reply:
So `orientation` is a data member of `struct Visitor`, it's defined after all the call operators have already been defined but still within `Visitor`. We are in fact modifying the passed-in `o` value, that is an `Orientation*` and it's being used to initialize `Orientation& orientation` because `Visitor` is initialized with `*o` when we say `Visitor{*o}`

You can think of this similar to a lambda capture of `[&orientation = *o]` which is also how a lambda works under the hood, it's defining an anonymous, function-local struct with a call operator whose data members are initialized to the capture list